Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/Ltc.azsli

```diff
@@ -189,42 +219,90 @@ void ClipQuadToHorizon(inout float3 p[5], out int vertexCount)
-// points are below the horizon), or a pentagon (when one point is below the horizon), or be a regular 4 point quad.
-void LtcClipAndNormalizeQuad(in float3 normal, in float3 dirToView, float3x3 ltcMat, inout float3 p[5], out int vertexCount)
+// Applies the LTC matrix to the clipped points of a quad.
+void ApplyLtcMatrixToQuad(in float3x3 ltcMat, inout float3 p[5], in int vertexCount)
 {
-    // Rotate ltc matrix
-    ltcMat = mul(ltcMat, BuildViewAlignedOrthonormalBasis(normal, dirToView));
-
     // Transform points into ltc space
-    p[0] = mul(ltcMat, p[0].xyz);
-    p[1] = mul(ltcMat, p[1].xyz);
-    p[2] = mul(ltcMat, p[2].xyz);
-    p[3] = mul(ltcMat, p[3].xyz);
-
-    ClipQuadToHorizon(p, vertexCount);
+    p[0] = mul(ltcMat, p[0]);
+    p[1] = mul(ltcMat, p[1]);
+    p[2] = mul(ltcMat, p[2]);
 
-    // visibility check
-    if (vertexCount == 0)
+    if (vertexCount > 3)
     {
-        return;
+        p[3] = mul(ltcMat, p[3]);
+    }
+    if (vertexCount > 4)
+    {
+        p[4] = mul(ltcMat, p[4]);
     }
+}
 
-    // project onto sphere
+// Projects the clipped points of a quad onto the sphere.
+void NormalizeQuadPoints(inout float3 p[5], in int vertexCount)
+{
+    // project quad points onto a sphere.
     p[0] = normalize(p[0]);
     p[1] = normalize(p[1]);
     p[2] = normalize(p[2]);
-    p[3] = normalize(p[3]);
-    p[4] = normalize(p[4]);
+
+    if (vertexCount > 3)
+    {
+        p[3] = normalize(p[3]);
+    }
+    if (vertexCount > 4)
+    {
+        p[4] = normalize(p[4]);
+    }
+}
+
+// Transforms the 4 points of a quad into the hemisphere of the normal
+void TransformQuadToOrthonormalBasis(in float3 normal, in float3 dirToView, inout float3 p[4])
+{
+    float3x3 orthoNormalBasis = BuildViewAlignedOrthonormalBasis(normal, dirToView);
+
+    // Transform points into orthonormal space
+    p[0] = mul(orthoNormalBasis, p[0]);
+    p[1] = mul(orthoNormalBasis, p[1]);
+    p[2] = mul(orthoNormalBasis, p[2]);
+    p[3] = mul(orthoNormalBasis, p[3]);
 }
 
-float IntegrateQuad(in float3 v[5], in float vertexCount, in bool doubleSided)
+// Integrates the edges of a quad for lambertian diffuse contribution.
+float IntegrateQuadDiffuse(in float3 v[5], in float vertexCount, in bool doubleSided)
 {
-    // Integrate
     float sum = 0.0;
 
+    NormalizeQuadPoints(v, vertexCount);
+
+    // There must be at least 3 points so don't check for the first 2 edges.
+    sum += IntegrateEdgeDiffuse(v[0], v[1]);
+    sum += IntegrateEdgeDiffuse(v[1], v[2]);
+
+    if (vertexCount > 3)
+    {
+        sum += IntegrateEdgeDiffuse(v[2], v[3]);
+        if (vertexCount == 5)
+        {
+            sum += IntegrateEdgeDiffuse(v[3], v[4]);
+        }
+    }
+
+    // Close the polygon
+    sum += IntegrateEdgeDiffuse(v[vertexCount - 1], v[0]);
+
+    // Note: negated due to winding order
+    sum = doubleSided ? abs(sum) : max(0.0, -sum);
```

Comment:
Why not simplify (and avoid future problems) by using sum = abs(sum)?

Reply:
I'm not sure what you mean. sum = abs(sum) is only for the double-sided case. Otherwise you don't want the back of the quad contributing.

Reply:
Scrap that - in the reverse case of negative you should get 0 and not abs(sum)